google/cloud/storage/testing/retry_tests.h

```diff
@@ -320,6 +320,243 @@ void PermanentFailureTest(Client& client,
+ * @tparam F a formal parameter, the googlemock representation of the mocked
+ *     call.
+ *
+ * @param oncall The internal type returned by EXPECT_CALL(...).
+ * @param tested_operation a function wrapping the operation to be tested.
+ * @param api_name the name of the api
+ * @return a function that implements the test.
+ */
+template <typename ReturnType, typename F>
+void TooManyFailuresStatusTest(
+    std::shared_ptr<testing::MockClient> const& mock,
+    ::testing::internal::TypedExpectation<F>& oncall,
+    std::function<Status(Client& client)> const& tested_operation,
+    char const* api_name) {
+  using canonical_errors::TransientError;
+  using ::testing::HasSubstr;
+  using ::testing::Return;
+  // A storage::Client with a simple to test policy.
+  Client client{std::shared_ptr<internal::RawClient>(mock),
+                LimitedErrorCountRetryPolicy(2)};
+
+  // Expect exactly 3 calls before the retry policy is exhausted and an error
+  // status is returned.
+  oncall.WillOnce(Return(StatusOr<ReturnType>(TransientError())))
+      .WillOnce(Return(StatusOr<ReturnType>(TransientError())))
+      .WillOnce(Return(StatusOr<ReturnType>(TransientError())));
+
+  Status status = tested_operation(client);
+  EXPECT_EQ(TransientError().status_code(), status.status_code());
+  EXPECT_THAT(status.error_message(), HasSubstr("Retry policy exhausted"));
+  EXPECT_THAT(status.error_message(), HasSubstr(api_name));
+}
+
+/**
+ * Tests that non-idempotent operations are *not* retried case for a `Client::*`
+ * member function.
+ *
+ * We need to verify that non-idempotent operations are not retied when the
+ * policy says so. The tests are quite repetitive, and all have the same
+ * structure:
+ *
+ * - Create a `storage::Client` with the right idempotency policies.
+ * - Setup the mock to return the right number of transient failures.
+ * - Call the API.
+ * - Verify an error status is returned, with the right contents.
+ *
+ * This function implements these tests, saving us a lot of repetitive code.
+ *
+ * @tparam ReturnType The low-level RawClient return type (modulo StatusOr).
+ * @tparam F a formal parameter, the googlemock representation of the mocked
+ *     call.
+ *
+ * @param oncall The internal type returned by EXPECT_CALL(...).
+ * @param tested_operation a function wrapping the operation to be tested.
+ * @param api_name the name of the api
+ * @param has_will_repeatedly `.WillRepeatedly()` can be called only once
+ *     in @p oncall. If this flag is set, the test will not setup that
+ *     expectation again.
+ * @return a function that implements the test.
+ */
+template <typename ReturnType, typename F>
+void NonIdempotentFailuresStatusTest(
+    std::shared_ptr<testing::MockClient> const& mock,
+    ::testing::internal::TypedExpectation<F>& oncall,
+    std::function<Status(Client& client)> const& tested_operation,
+    char const* api_name, bool has_will_repeatedly = false) {
```

Comment:
`has_will_repeatedly` unused?

Reply:
Indeed, fixed.

You probably noticed, but that was a leftover from using `EXPECT_DEATH_IF_SUPPORTED()` with `-fno-exceptions`, one cannot check side-effects with `EXPECT_DEATH*`.